storage: internal documentation for ChangeReplicas

petermattis opened the pull request:
Fixes #7709.

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/7822)

<!-- Reviewable:end -->

petermattis committed:
storage: internal documentation for ChangeReplicas

Fixes #7709.

petermattis commented:
A work in progress, but worthy of comments.

bdarnell commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 0 of 1 files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[storage/replica_command.go, line 2869 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7822#-KMbfeBK_HcvLfPLyd4t:-KMbfeBLhXnHcl7tOeBA:-2098017891) ([raw file](https://github.com/cockroachdb/cockroach/blob/35861229c246758a547c0b426677ff733adddb72/storage/replica_command.go#L2869)):*
> ```Go
> // holder. defaultProposeRaftCommandLocked notices that the EndTransaction
> // contains a ChangeReplicasTrigger and proposes a ConfChange to Raft (via
> // raft.RawNode.ProposeConfChange). When this phase completes, the replica
> ```

This is out of order; the raft leader doesn't start sending heartbeats (etc) until handleRaftReady calls ApplyConfChange, after the EndTransaction has been applied. (so maybe it's a fourth phase?)

---

*[storage/replica_command.go, line 2881 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7822#-KMbg5f7Dqrk3gaE6369:-KMbg5f7Dqrk3gaE636A:-1254940529) ([raw file](https://github.com/cockroachdb/cockroach/blob/35861229c246758a547c0b426677ff733adddb72/storage/replica_command.go#L2881)):*
> ```Go
> //
> // TODO(peter): Why does a removed replica receive the EndTransaction
> // containing the ChangeReplicasTrigger? Is this accidental or guaranteed by
> ```

It's accidental. The replica isn't removed until after the leader has applied the EndTransaction (and therefore after that message has been appended to a quorum of logs), so if it is up at the time of its removal it will likely get this message. However, if it's not up or the message is dropped for any reason, it will not be retried and the replica will be left for the GC queue to eventually clean up.

---

*[storage/replica_command.go, line 2902 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7822#-KMbgdMYgSiRVoF42UHE:-KMbgdMYgSiRVoF42UHF:932425347) ([raw file](https://github.com/cockroachdb/cockroach/blob/35861229c246758a547c0b426677ff733adddb72/storage/replica_command.go#L2902)):*
> ```Go
> // applied.
> //
> // TODO(peter): What happens if the replica is brought up to date via Raft log
> ```

Currently the reservation is left hanging until it expires. This is a somewhat unlikely scenario (it requires that the replica GC queue hasn't removed the replica, *and* that the raft log GC hasn't removed log entries that would be required for this), but it's possible. I outlined one possible scenario in #7800.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7822#-:-KMbhTp_5-Rbt1ssJ8fV:1417088721)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: 0 of 1 files reviewed at latest revision, 3 unresolved discussions.

---

*[storage/replica_command.go, line 2869 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7822#-KMbfeBK_HcvLfPLyd4t:-KMjXWuCiTm6yi95FxAA:-1304847574) ([raw file](https://github.com/cockroachdb/cockroach/blob/35861229c246758a547c0b426677ff733adddb72/storage/replica_command.go#L2869)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> This is out of order; the raft leader doesn't start sending heartbeats (etc) until handleRaftReady calls ApplyConfChange, after the EndTransaction has been applied. (so maybe it's a fourth phase?)
</details>
Adjusted this section. Expanded to 4 phases.

---

*[storage/replica_command.go, line 2881 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7822#-KMbg5f7Dqrk3gaE6369:-KMjY97ku73o6EKComeZ:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/35861229c246758a547c0b426677ff733adddb72/storage/replica_command.go#L2881)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> It's accidental. The replica isn't removed until after the leader has applied the EndTransaction (and therefore after that message has been appended to a quorum of logs), so if it is up at the time of its removal it will likely get this message. However, if it's not up or the message is dropped for any reason, it will not be retried and the replica will be left for the GC queue to eventually clean up.
</details>
Done.

---

*[storage/replica_command.go, line 2902 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7822#-KMbgdMYgSiRVoF42UHE:-KMjYToEYzeBojgTpL9R:157125279) ([raw file](https://github.com/cockroachdb/cockroach/blob/35861229c246758a547c0b426677ff733adddb72/storage/replica_command.go#L2902)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> Currently the reservation is left hanging until it expires. This is a somewhat unlikely scenario (it requires that the replica GC queue hasn't removed the replica, *and* that the raft log GC hasn't removed log entries that would be required for this), but it's possible. I outlined one possible scenario in #7800.
</details>
I added a link to #7849.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7822)*
<!-- Sent from Reviewable.io -->
